Add @inbounds annotations to any(), all() and count()

nalimilan opened the pull request:
Beside removing unnecessary bounds checks, `@inbounds` allows the compiler
to optimize away the loop when method is constant for input type, e.g.
`all(x -> true, X)` or `all(isnull, 1:3)`.  An optimized method already existed
for `count(::Any, ::AbstractArray)`, remove it since it is redundant with
method for the general iterables after adding `@inbounds`.

Fixes https://github.com/JuliaLang/julia/issues/21256.

I'm still not completely sure what our policy is regarding `@inbounds`: if it's OK to use for `AbstractArray` when working with an index obtained via `eachindex`, or equivalently using the iteration protocol, why wouldn't it be OK for any iterable? It seems a bit silly to duplicate all methods. Ideally, `for x in X` would automatically imply `@inbounds` (at least if the type implementation decides to do so). Cf. https://github.com/JuliaLang/julia/issues/20469.

Also I don't know how to test this, except with benchmarks?

vchuravy commented:
Remember to quote macros like `@inbounds` otherwise the respective Github user will be pinged.

nalimilan commented:
> Remember to quote macros like @inbounds otherwise the respective Github user will be pinged.

Right, this one came from the commit message and I forgot to change it.

ChrisRackauckas commented on base/reduce.jl:
```diff
@@ -693,10 +714,11 @@ function count(pred, itr)
     end
     return n
 end
+# @inbounds allows the compiler to optimize out the loop in case eq is constant
```
>in case eq is constant

in case pred is constant

nalimilan replied:
Thanks, fixed.

TotalVerb commented on base/reduce.jl:
```diff
@@ -575,6 +575,13 @@ function any(f, itr)
     return false
 end
+# @inbounds allows the compiler to optimize out the loop in case f is constant
+function any(f, a::AbstractArray)
```
Is our policy that it's OK for `AbstractArray` for indices from `eachindex`? A more conservative version would only use `Array`, which will be a strict performance change instead of a possibly breaking one (albeit, only broken code will be broken).

nalimilan replied:
See #20469. In short, in the long term this will be safe, but for now it may have unintended effects for some custom types. Since this will only be a problem if they implement the iteration protocol incorrectly, I would say it's OK as a temporary situation, but feel free to argue otherwise on the other issue.

nalimilan commented:
This shouldn't be merged until we have clearly decided a policy for `@inbounds`: it would be good to avoid duplicating the methods.

nalimilan committed:
Add @inbounds annotations to any() and all()

Beside removing unnecessary bounds checks, @inbounds allows the compiler
to optimize away the loop when method is constant for input type, e.g.
all(x -> true, X) or all(isnull, 1:3). An optimized method already existed
for count(::Any, ::AbstractArray), remove it since it is redundant with
method for the general iterables after adding @inbounds.

nalimilan commented:
According to https://github.com/JuliaLang/julia/issues/20469#issuecomment-327827021, it is OK to add `@inbounds` annotations to methods accepting any iterable, since by using `@boundscheck` annotations the implementation guarantees indices are always valid. So I've updated the PR to simply add the needed annotations to the generic methods.

@mbauman Is that right?

nalimilan commented:
@mbauman Bump! :-)

mbauman commented on base/reduce.jl:
```diff
@@ -658,16 +660,10 @@ julia> count([true, false, true, true])
 function count(pred, itr)
     n = 0
-    for x in itr
+    # @inbounds allows the compiler to optimize out the loop in case f is constant
```
Copy-paste comment typo.  `f` -> `pred` in this case.

mbauman commented:
Yes, I think this is reasonable.  The only thing that gives me pause is that if we're going to do this, maybe we should bite the bullet and just do it everywhere.  That is, let's just do it everywhere — something like https://github.com/JuliaLang/julia/pull/15291.

nalimilan commented:
Indeed, if that's now possible it would really be nice. How hard would that be, and in what timeframe?

yuyichao reviewed (changes requested):
Even if we really want to allow this for arbitrary iterable/user defined types, which I think is a bad idea, we still shouldn't apply that on the callbacks.

mbauman commented:
Ah, great point.  Isn't that an even _better_ reason to do this in lowering, then?  We don't have a syntax that allows us to do `for @inbounds x in y`.

As far as how hard it is, it just takes someone with a modicum of lisp-foo to add something like `` `(macrocall @inbounds (call next …))``.  My quick stab at it failed.

yuyichao commented:
I agree and I also think https://github.com/JuliaLang/julia/pull/15291 is a better approach. Property following the iteration protocal is still a different concept than inbounds.

yuyichao commented:
And AFAICT as long as we agree to introduce `unsafe_next` (and whatever other function it'll be called if we decide to change the iteration protocol), fixing that PR shouldn't be very hard.

andyferris commented:
I vote for fixing up and merging #15291, for a much more general fix here. Otherwise we'd be searching for these cases for ever, and getting major problems like segfaults when prototyping new types that iterate and making silly mistakes.

andyferris commented:
> `f(x) = any(ismissing, x); f([1])` are now resolved statically to false.

Whoa - that's amazing! Mind blown.
